### PR TITLE
Fixed issue 3770

### LINKED ISSUE #3770: [ENG-159] Language popup shown off-screen
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

When the language selector is in the lower part of the screen, the menu pops up partly or entirely outside the browser screen

### Expected Behavior

The whole popup menu should be visible

### Steps To Reproduce

Click the language selector when it is in the lower part of the screen

### What platform(s) does this occur on?

Web

### What browser(s) does this occur on?

Chrome

### Anything else?

https://github.com/lensterxyz/lenster/assets/667227/1398d7ea-d14b-421a-bfb4-093575560419

<sub>[ENG-159](https://linear.app/heyxyz/issue/ENG-159)</sub>

### PR DESCRIPTION
**Pull request by prateekkalra**
## What does this PR do?

Makes the placement of locale dropdown dynamic. 
This PR handles overflow from all 4 directions(Top, bottom,left,right).
The position of dropdown even flips(bottom to top) when the user is scrolling while the dropdown is open.

PS: The other PR doesn't handle the above cases.

/claim #3770

## Related issues

Fixes #3770

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Enhancement (non-breaking small changes to existing functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### <samp>🤖 Generated by Copilot at f0d4eb8</samp>

*  Add `@popperjs/core` dependency to `apps/web/package.json` and `pnpm-lock.yaml` files to use popper library for positioning UI elements ([link](https://github.com/heyxyz/hey/pull/3894/files?diff=unified&w=0#diff-14b60f636e1a2b0061da57aaf231cb1ed15a5dc0c592425ed82e58fec95d42d8R63), [link](https://github.com/heyxyz/hey/pull/3894/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR153-R155), [link](https://github.com/heyxyz/hey/pull/3894/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR5307-R5310))
*  Remove extra newline at the end of `apps/web/package.json` file for formatting consistency ([link](https://github.com/heyxyz/hey/pull/3894/files?diff=unified&w=0#diff-14b60f636e1a2b0061da57aaf231cb1ed15a5dc0c592425ed82e58fec95d42d8L101-L100))
*  Create custom hook `usePopper` in `apps/web/src/hooks/usePopper.tsx` file to wrap popper library and return references for trigger and container elements ([link](https://github.com/heyxyz/hey/pull/3894/files?diff=unified&w=0#diff-e428af1ffc333bfba1b69786335f8c457cc77d2bff0326847d8c2792e515eb73L1-R47))
*  Import and use `usePopper` hook in `apps/web/src/components/Shared/Footer/Locale.tsx` file to create and link popper instance for locale selector menu ([link](https://github.com/heyxyz/hey/pull/3894/files?diff=unified&w=0#diff-2d9af24d277e24d0e5b89b5e09bf7745dd1a4c55b9630b816484d0319313cacbR13), [link](https://github.com/heyxyz/hey/pull/3894/files?diff=unified&w=0#diff-2d9af24d277e24d0e5b89b5e09bf7745dd1a4c55b9630b816484d0319313cacbR32-R43), [link](https://github.com/heyxyz/hey/pull/3894/files?diff=unified&w=0#diff-2d9af24d277e24d0e5b89b5e09bf7745dd1a4c55b9630b816484d0319313cacbL36-R81))

## Emoji

🎣🌐🔧

### CODE REVIEW

**Comment by bigint** [comment]
@prateekkalra dont headless have this native implementation?

**Comment by prateekkalra** [comment]
@bigint @headlessui/react doesn't have a native implementation. @headlessui-float/react has an implemention, but it is having conflicts with the current version of @headlessui/react in the project.
I think using popperjs is the way to go.
Refer https://github.com/tailwindlabs/headlessui/issues/30 for more details.

**Comment by bigint** [comment]
But this still works in popover and not only in menu!

**Comment by prateekkalra** [comment]
Maybe because popovers are meant to be dynamically positioned but menu  isn't. Could you please tell me what is wrong with this solution? The menu works as expected.

File: apps/web/src/components/Shared/Footer/Locale.tsx [hunk 1]
```diff
@@ -28,41 +29,80 @@ const Locale: FC = () => {
...
+              });
+            }
+          }
+        }
+      },
+      {
+        name: 'preventOverflow'
+      }
+    ]
+  });
```
Comment (bigint) [inline apps/web/src/components/Shared/Footer/Locale.tsx:66]: we have 10+ menus in the app, we need to make it reusable and small

**Comment by bigint** [comment]
Looks like #3893 approach looks better, so closing this 🙇🏼

File: apps/web/src/components/Shared/Footer/Locale.tsx (hunk 1 above)
Comment (BilalShah3) [reply apps/web/src/components/Shared/Footer/Locale.tsx:66]: but this hook did not handle the situation when we are using with portal or teleport then floating element will not acnhor to refrence element example when scrolling . can you also add this.

**Pull request closed without merging**